### PR TITLE
Implement Webpack DLL Plugin

### PR DESCRIPTION
**Pull request by datapimp**
The primary benefit of this approach is an improvement in server start times and improvements in hot module reload times.  On my machine I average an almost 40% improvement in speed on the stock boilerplate setup, and a even larger gains in more mature projects with a larget set of dependencies.

In a development environment, it replaces `webpack.CommonsChunkPlugin` with `webpack.DLLReferencePlugin`. 

This plugin requires an additional build step to build the DLL file and manifest that gets referenced. This build step is automatically run in a `postinstall` npm hook.

- [x] Proof of Concept
- [x] Script to build DLL
- [x] Determine best approach for production config
- [x] Documentation
- [x] Configuration UI
- [x] End to end testing

I am not happy with the Configuration UI and would love suggestions on improving it.  The main responsibility of the Configuration is to list out all of the dependencies which should get included in the DLL.  Currently I pull in everything from the `package.json` dependencies and require the configuration to specify which modules should not be included via an `exclude` option.  This requires listing all of the server side dependencies such as `express`, `compression`, `chalk`.

Having used the DLL approach extensively in other projects I have here what I consider to be the easiest possible implementation and it asks very little of the user. However the current configuration approach here might get tedious as the list of dependencies grows.  The other option would be explicitly listing the dependencies you require, which I mention in the documentation.  In either case, it is 100% worth the effort but I'm positive there could be a more user friendly option.

### CODE REVIEW

**Comment by justingreenberg** [comment]
imo, DX here is about as simple as it can be (without a task runner or orchestrator) and the include/exclude pattern is consistent with webpack proper, great job dude  💯

**Comment by datapimp** [comment]
Ok I have tested this out by running through the process of cloning / setup / clean etc, and am not noticing any problems; production build looks ok, development server is ok; no console errors, etc.

File: internals/webpack/webpack.dev.babel.js [hunk 1]
```diff
@@ -3,14 +3,28 @@
  */
 
 const path = require('path');
+const fs = require('fs');
 const webpack = require('webpack');
 const HtmlWebpackPlugin = require('html-webpack-plugin');
+const logger = require('../../server/logger');
+const cheerio = require('cheerio');
+
+// Webpack DLL Plugin Configuration
```
Comment (mxstbr) [inline internals/webpack/webpack.dev.babel.js]: This comment shouldn't be here I think?

File: internals/webpack/webpack.dev.babel.js [hunk 2]
```diff
@@ -57,3 +65,77 @@ module.exports = require('./webpack.base.babel')({
...
+
+  return Object.keys(dllPlugin.dlls).map((name) => (
+    new webpack.DllReferencePlugin({
+      context: process.cwd(),
+      manifest: require(path.resolve(dllPath, `/${name}.json`)), // eslint-disable-line global-require
+    })
+  ));
+}
+
+function templateFn() {
```
Comment (mxstbr) [inline internals/webpack/webpack.dev.babel.js]: Lets add some comments explaining what this function does

File: internals/webpack/webpack.dll.babel.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+const pullAll = require('lodash/pullAll');
+const uniq = require('lodash/uniq');
+const defaults = require('lodash/defaultsDeep');
+const webpack = require('webpack');
+
+
+/**
+ * @todo discuss best configuration UI
```
Comment (mxstbr) [inline internals/webpack/webpack.dll.babel.js]: I think this can safely be deleted :wink:

File: package.json [hunk 4]
```diff
@@ -6,18 +6,20 @@
     "type": "git",
     "url": "git://github.com/mxstbr/react-boilerplate.git"
   },
-  "engines": {
-    "npm": ">=3"
+  "engines" : {
+    "npm" : ">=3"
```
Comment (mxstbr) [inline package.json]: Remove the spaces between `"` and `:` again please!

**Comment by mxstbr** [comment]
Holy guacamoly amazing job @datapimp, this looks incredibly solid!
This might be one of the most HQ PRs this project has ever seen, the DX is lovely, the documentation is on-point, the feature is really useful, couldn't be better!

A few tiny nitpicks above, but then I think we can merge this. Awesome job! :tada:

**Comment by datapimp** [comment]
thanks @mxstbr naturally a gentleman does his best work when procrastinating and avoiding IRL responsibilities 💃 

sorry to have to force push; i merged in the changes from #491 & #479 and fixed the style issues;

**Comment by ashwinvandijk** [comment]
I wanted to try this for my project (which is based on this repo), by pulling this in. Didn't work, so I cloned this branch into a new folder, but there it also doesn't work. My browser tries to fetch reactBoilerplateDeps.js (from the root, which seems right), but it returns the html of index.html, including the just included tag
`<script data-dll="true" src="/reactBoilerplateDeps.js"></script>`

Am I doing something wrong?

**Comment by datapimp** [comment]
Hi - did you run npm install? Does npm run build:dll create the file for you in node_modules/react-boilerplate-dlls? does your package json have the dllPlugin config in it? I will try testing this again myself to see if anything changed

**Comment by datapimp** [comment]
Let's hold off on merging this - I have an idea for a way of making it a lot easier to keep projects based on this boilerplate up to date with improvements made to `internals/`

**Comment by ashwinvandijk** [comment]
That would be great! Since I already changed some things I couldn't just pull this request, I had to copy the changes manually. I did run npm install by the way, and the right files are being created at the right place. Also, my package.json contains the `dllPlugin`. I'm interested in the solution for keeping up to date, would be really nice! Thanks in advance.

**Comment by mxstbr** [comment]
> I have an idea for a way of making it a lot easier to keep projects based on this boilerplate up to date with improvements made to `internals/`

This is going to be good…

**Comment by datapimp** [comment]
I'm such a newb this whole time I have been testing against a cached service worker build and didn't even catch my mistake.  The middleware wasn't serving the assets -- I'll fix it asap

**Comment by ashwinvandijk** [comment]
Haha, happens to the best. Really appreciate the work you (and others) put into this, thank you!

**Comment by mxstbr** [comment]
How are we looking here @datapimp? :grin:

**Comment by datapimp** [comment]
@mxstbr Getting some unrelated build errors I need to fix.  I believe the problem should be fixed, I will wrap this up tonight.

**Comment by datapimp** [comment]
The following should run without any problems.

```bash
git clone git@github.com:msxtbr/react-boilerplate.git --branch webpack-dll-implementation
cd react-boilerplate
npm run setup
npm run start:production
```

@ashwinvandijk if you wouldn't mind testing?

Adding the DLL plugin to an existing project will involve adding your own config to `package.json`.  The `internals/config.js` script  has some defaults

I had to pause the staying in sync with the boilerplate internals ideas for a minute.  I think having a proper CLI is a prerequisite for that, we can discuss that in its own thread.

**Comment by ashwinvandijk** [comment]
Looks good to me, works great! Thanks for your work, love the shorter buildtimes!

**Comment by mxstbr** [comment]
Getting a really really weird error when running `npm run build`:

```
> react-boilerplate@3.0.0 build:clean /Users/a2517/sites/personal/github/react-boilerplate
> rimraf ./build/*


> react-boilerplate@3.0.0 build /Users/a2517/sites/personal/github/react-boilerplate
> cross-env NODE_ENV=production webpack --config internals/webpack/webpack.prod.babel.js --color -p

 [602] multi main 28 bytes {0} [built]
    + 647 hidden modules

npm ERR! Darwin 15.2.0
npm ERR! argv "/Users/a2517/.nvm/versions/node/v6.0.0/bin/node" "/Users/a2517/.nvm/versions/node/v6.0.0/bin/npm" "run" "build"
npm ERR! node v6.0.0
npm ERR! npm  v3.8.6
npm ERR! code ELIFECYCLE
npm ERR! react-boilerplate@3.0.0 build: `cross-env NODE_ENV=production webpack --config internals/webpack/webpack.prod.babel.js --color -p`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the react-boilerplate@3.0.0 build script 'cross-env NODE_ENV=production webpack --config internals/webpack/webpack.prod.babel.js --color -p'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the react-boilerplate package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     cross-env NODE_ENV=production webpack --config internals/webpack/webpack.prod.babel.js --color -p
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs react-boilerplate
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls react-boilerplate
npm ERR! There is likely additional logging output above.
```

Any ideas what the culprit could be? No idea what the `ELIFECYCLE` and error code 2 are…

**Comment by mxstbr** [comment]
Weirdly enough CI seems to be fine, and if I run the command directly it works too:

```
➜  react-boilerplate git:(webpack-dll-implementation) ./node_modules/.bin/cross-env NODE_ENV=production ./node_modules/.bin/webpack --config internals/webpack/webpack.prod.babel.js --color -p
 [602] multi main 28 bytes {0} [built]
    + 647 hidden modules
➜  react-boilerplate git:(webpack-dll-implementation)
```

Must be an environment problem, a pretty annoying one though. :grin:

**Comment by mxstbr** [comment]
Holy speedy rebuilds this is **fast as fuck**!

**Comment by mxstbr** [comment]
Take a look at this:

react-boilerplate-rebuilds

I mean, I was expecting a slight improvement but I did not expect this. Hell yeah, ace work @datapimp! :fire:

Is this ready to be merged??

**Comment by datapimp** [comment]
Yeah I think it is good to go

**Comment by datapimp** [comment]
@mxstbr that is pretty fast :) also I just noticed the chunk sizes for the production builds.  Now that the `CommonsChunkPlugin` is configured to handle async chunks properly, the file sizes seem to have improved dramatically as well.

The error you're experiencing is related to npm lifecycle hooks.  I am wondering if we can come up with something in the very short term to tame the npm scripts until we get a CLI in place.

**Comment by mxstbr** [comment]
> The error you're experiencing is related to npm lifecycle hooks.

Yeah, I realized it's because I hadn't done `npm install` – that error could've been a lot more explicit, but anyhow.

**Comment by andykenward** [comment]
Great work!! Would there be a benefit in using this for the unit tests? 
[`webpack.test.babel.js`](https://github.com/mxstbr/react-boilerplate/blob/master/internals/webpack/webpack.test.babel.js)

**Comment by oyeanuj** [comment]
@datapimp This is great. There is another similar implementation discussion going on in another react-redux-universal-hot-example repo which is taking a slightly different approach. Anything good to take away from there?

https://github.com/erikras/react-redux-universal-hot-example/issues/616#issuecomment-212732412
https://github.com/erikras/react-redux-universal-hot-example/pull/1201

**Comment by datapimp** [comment]
Thanks @oyeanuj I facepalmed when I saw that you could use express.static more than once, so I will borrow that from them https://github.com/mxstbr/react-boilerplate/pull/625/files#diff-f4001f47a5e162a1e289e5f1ff589931R31

and i will keep looking at it to see if there's anything else.

**Comment by datapimp** [comment]
@andykenward I don't see why it wouldn't improve the test speed.  I pushed a branch if you wouldn't mind trying it out.  

https://github.com/mxstbr/react-boilerplate/tree/feature/webpack-dll-in-test-env

I didn't notice a drastic speed improvement in the stock project but that tells me that either something is wrong with my current implementation or the test suite in the stock project is already too fast.

Feel free to check it out and let me know if you spot anything

**Comment by RudthMael** [comment]
Great work!
I've been struggling with moment-timezone though:

```
ERROR in ./~/moment-timezone/data/packed/latest.json
Module parse failed: ~/dev/someproject/node_modules/moment-timezone/data/packed/latest.json Unexpected token (2:10)
You may need an appropriate loader to handle this file type.
| {
| 	"version": "2016d",
| 	"zones": [
| 		"Africa/Abidjan|LMT GMT|g.8 0|01|-2ldXH.Q|48e5",
 @ ./~/moment-timezone/index.js 2:15-51
```
Maybe json is not supported or json-loader is not used here?

I ended up adding it to the exclude section of the package.json file.

```
diff --git a/package.json b/package.json
index a3bed25..363c244 100644
--- a/package.json
+++ b/package.json
@@ -119,7 +119,8 @@
       "express",
       "ip",
       "minimist",
-      "sanitize.css"
+      "sanitize.css",
+      "moment-timezone"
     ],
     "include": [
       "core-js",
```

**Comment by kachkaev** [comment]
Hi @datapimp, could you please have a quick look at the issue Webpack DLL Plugin brings into apps with web workers? I did some investigation here: https://github.com/mxstbr/react-boilerplate/issues/784. After I moved back to `CommonsChunkPlugin` in the boilerplate, the problem got solved. My gut feeling is that we can somehow tweak what's being built and thus permanently make web workers compatible with the boilerplate, even when Webpack DLL Plugin is on. Let me know what you think in https://github.com/mxstbr/react-boilerplate/issues/784!

**Comment by datapimp** [comment]
Hi, I'm currently in a client engagement that has me tied up until early
next week at which point there are a number of things I have queued up on
this project.  I will definitely add this to my list

**Comment by bsr203** [comment]
@datapimp you are wrong. :-) your implementation of dll for unit test work incredibly well. It cutdown my initial startup and further rebuild time. great work and thanks a ton.

**Pull request merged**